Add PEP 561-style type information

pcorpet opened the pull request:
Note that I'm not sure how to correctly add the files in the package. The `*.pyi` files need to be installed, so as the empty `py.typed` file.

As for the types, we've been using them for a while, with `scrapy` types as well. They are not complete but it's a start, and there's no reason for those types to live in our repo when they can benefit others.

Gallaecio commented:
I was not aware of these `pyi` files, but could you instead include the typing hints into the actual files? That’s the way we are doing it in Scrapy, and I think it makes things easier (e.g. if you change a function you only need to edit its file, not also the corresponding type file).

pcorpet commented:
I'm happy to go that way as it's indeed easier to maintain.

Note that it enforces to use `python 3`, and I have no clue whether `parsel` still support `python 2`.

Gallaecio commented:
You are right, [we still support Python 2](https://github.com/scrapy/parsel/blob/master/setup.py#L69) :facepalm:.

Maybe it’s time for us to “fix” that :slightly_smiling_face:

Gallaecio commented:
Fixed, please merge/rebase.

pcorpet commented:
Done. Note that the annotations I've added are not enough to type `parsel` entirely, it's only typing the public API so that other typed code can depend on it properly.

Gallaecio commented:
> Note that the annotations I've added are not enough to type `parsel` entirely, it's only typing the public API so that other typed code can depend on it properly.

That’s perfectly fine. It’s definitely an improvement over not having any typing hints at all.

Could you also add mypy to the CI? It should only require editing `tox.ini` and `.github/workflows/checks.yml`. See similar files in https://github.com/scrapy/itemadapter for reference.

pcorpet commented:
Alright, I've added `mypy`. Few things:
- I've strongly typed the selector tests, so that we can check most of the API
- I've added an extra typing test file that doesn't run any test but only check for correct and incorrect typing
- I've had to update some tests (especially the ones that were checking `isinstance`, but also the one that extends the classes): that is intended

Gallaecio commented on parsel/selector.py:
```diff
@@ -290,7 +325,18 @@ def re(self, regex, replace_entities=True):
         return extract_regex(regex, self.get(), replace_entities=replace_entities)
 
-    def re_first(self, regex, default=None, replace_entities=True):
+    @typing.overload
+    def re_first(self, regex: Union[str, Pattern[str]], default: None = None,
+                 replace_entities: bool = True) -> Optional[str]:
+        pass
```
I see Codecov is complaining about these. See https://github.com/nedbat/coveragepy/issues/970#issuecomment-612602180

pcorpet replied:
Good point. Done.

pcorpet commented on parsel/selector.py:
```diff
@@ -52,20 +57,29 @@ def create_root_node(text, parser_cls, base_url=None):
+    def __getitem__(self, pos: slice) -> 'SelectorList[_SelectorType]':
+        pass
+
+    def __getitem__(self, pos: Union[int, slice]) -> Union[_SelectorType, 'SelectorList[_SelectorType]']:
```
It's somehow superfluous only: it's useless for callers of the function, however to type the internals of the function, it's better to specify the types. Note that `mypy` checks that they are compatible. So except you're against it, I'd like to keep it.

pcorpet commented on parsel/selector.py:
```diff
@@ -52,20 +57,29 @@ def create_root_node(text, parser_cls, base_url=None):
 
-    def xpath(self, xpath, namespaces=None, **kwargs):
+    def xpath(self, xpath: str, namespaces: Optional[Mapping[str, str]] = None, **kwargs) \
+            -> 'SelectorList[_SelectorType]':
```
Yep, we should switch parsel to use `black` soon :-) Done.

Gallaecio reviewed (commented):
By the way, in pull requests, once you got some feedback, it is often better to keep sending commits instead of squashing and force-pushing new changes. Squashing makes review harder on those who already reviewed your previous changes, because instead of reviewing your new changes they need to review the whole change set again (at least modified files).

If you are worried about keeping a clean commit history, mind that we can easily squash your changes before merging, so that should not be a concern.

pcorpet commented on parsel/__init__.py:
```diff
@@ -9,6 +9,13 @@
-from parsel import xpathfuncs # NOQA
+from parsel import xpathfuncs  # NOQA
+
+__all__ = [
+    'Selector',
+    'SelectorList',
+    'css2xpath',
+    'xpathfuncs',
+]
```
This is to mention to mypy that the import above are actually meant to be be re-exported. see https://mypy.readthedocs.io/en/stable/command_line.html#cmdoption-mypy-no-implicit-reexport

Gallaecio replied:
Oh, I see. What about adding `--implicit-reexport` to the `mypy` call in `tox.ini` instead? I don’t think this is something we want to check for, at least at the moment.

pcorpet replied:
This would work for us, but would be problematic for users of the library. Any users running mypy on code that relies on `parsel` would then need to add the flag as well.

Gallaecio replied:
I must say I don’t like that mypy is forcing us to make this style change just to not break stuff for other people. But given it’s a trivial change, I guess we can budge.

In that case, I would only suggest to place `__all__` before imports, to [comply with PEP8](https://www.python.org/dev/peps/pep-0008/#module-level-dunder-names).

pcorpet replied:
Done.
